Add imagic to community pipelines

MarkRich opened the pull request:
Part of https://github.com/huggingface/diffusers/issues/895 and bigger story https://github.com/huggingface/diffusers/issues/841

Followed the rough code / parameters given here: https://github.com/justinpinkney/stable-diffusion/blob/main/notebooks/imagic.ipynb.

A few notes for reviews: 
* Seems like a bad idea to modify the stable diffusion pipeline to take text embeddings; that said it wasn't clear how we plan to re-use this pipeline given these custom generated embeddings. Any advice on how to pass these in? Or is modification given reasonable? 
* Also seems like a bad idea to define the learning rates where I do. That said this is a bit of an odd stable diffusion pipeline since we're training in the `__call__`, so not sure where these are expected to go.
* Similar comment as in https://github.com/huggingface/diffusers/pull/951: Test condition given in readme doesn't work locally and need to make a few changes to make it work locally, but I expect this is only an issue w/ local testing and/or I am missing something w.r.t. how these scripts are intended to be tested. 

## Results: 

Requires 24gb of vram and takes about 7-10 minutes on a 3090, though apparently it's 30g vram in 5pm on an a100 in original script. So reasonable performance? 

**Initial Image**: 
![obama](https://user-images.githubusercontent.com/2020546/197438440-fd321a5b-885d-49a5-b6ba-e6f7bcd24aeb.png)

Prompt: "A photo of Barack Obama smiling with a big grin"

**Image from just text embedding:** 
![image_from_pure_text_embeds](https://user-images.githubusercontent.com/2020546/197438508-1ebc1750-52ab-4522-b3cf-70b77f858f4f.png)

**Image after text embeddings have been optimized:**
![image_from_modified_text_embeds](https://user-images.githubusercontent.com/2020546/197438535-2dcc805f-31ed-417e-ae89-bc9dff0b8d43.png)

**Final image at alpha = 0.8**
![final_image_finetuned_and_embedding_8](https://user-images.githubusercontent.com/2020546/197438564-74afbc47-e5ea-4396-955d-20a1d2d541cc.png)

**Final Image at alpha = 1.5**
![final_image_finetuned_and_embedding_15](https://user-images.githubusercontent.com/2020546/197438612-ffd5f230-00db-4116-b6a2-fed32962ee4e.png)

**Final Image at alpha = 2.**
![final_image_finetuned_and_embedding_20](https://user-images.githubusercontent.com/2020546/197438650-c92fcdb2-b619-45a0-b99f-e95017267cab.png)

Looking forward to any comments!

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/diffusers/pr_958). All of your documentation changes will be reflected on that endpoint.

patrickvonplaten reviewed (commented):
Great job @MarkRich!

The design is generally fine with me! Also related to https://github.com/huggingface/diffusers/issues/955 - seems like there are multiple use cases for custom `text_embeddings` already

@patil-suraj could you do a more in-depth review?

patil-suraj commented on examples/community/imagic_stable_diffusion.py:
```diff
@@ -0,0 +1,350 @@
+        freeze_params(text_embeddings)
+
+        # Now we fine tune the unet to better reconstruct the image
+        unfreeze_params(self.unet.parameters())
```
let's use `requires_grad_` here

patil-suraj commented on examples/community/imagic_stable_diffusion.py:
```diff
@@ -0,0 +1,350 @@
+        # Now we fine tune the unet to better reconstruct the image
+        unfreeze_params(self.unet.parameters())
+        self.unet.train()
+        optimizer = torch.optim.Adam(
+            self.unet.parameters(),  # only optimize unet
+            lr=diffusion_model_learning_rate,
+        )
```
same as above, maybe add an option to use 8 bit adam. Could be done in follow up PR

MarkRich replied:
will do so in a follow up PR, I've been running into this issue: https://github.com/huggingface/diffusers/issues/946 so may be a bit more complicated for me to get running near term.

patil-suraj commented on examples/community/imagic_stable_diffusion.py:
```diff
@@ -0,0 +1,350 @@
+        )
+        progress_bar = tqdm(range(model_fine_tuning_optimization_steps), disable=not accelerator.is_local_main_process)
+
+        print("Next fine tuning the entire model to better reconstruct the init image")
```
Let's use `logger` here

patil-suraj commented on examples/community/imagic_stable_diffusion.py:
```diff
@@ -0,0 +1,350 @@
+        new_text_embeddings = alpha * text_embeddings_orig + (1 - alpha) * text_embeddings
+        image = pipeline(
+            prompt, text_embeddings=new_text_embeddings, scale=7.5, num_inference_steps=num_inference_steps
+        ).images[
+            0
+        ]  # , latents=noise_latents).images[0]
```
```suggestion
        ).images[0]
```

patil-suraj commented on src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py:
```diff
@@ -226,7 +227,9 @@ def __call__(
             )
             text_input_ids = text_input_ids[:, : self.tokenizer.model_max_length]
-        text_embeddings = self.text_encoder(text_input_ids.to(self.device))[0]
+
+        if text_embeddings is None:
+            text_embeddings = self.text_encoder(text_input_ids.to(self.device))[0]
```
I'm still not sure if we should do this, we are discussing it here #955

patil-suraj reviewed (commented):
Very cool @MarkRich , thanks a lot for adding the feature

Th pr looking really good! Just left a few nits. 

And I'm not sure yet, if we should modify the `StableDiffusionPipeline` to allow text_embeddinsg, we are discussing it here #955

For now, since we are adding a custom pipeline, I would suggest we could add to functions to the pipeline.

1. `pipeline.train` to train the embeddings and mode
2. `pipeline.__call__` or `pipeline.generate` to generate the images.

wdyt @patrickvonplaten

MarkRich commented:
Addressed all your comments @patil-suraj aside from 8-bit optimization which may take slightly longer for to instrument due to an unrelated error. Let me know if you have any other comments!

patil-suraj reviewed (approved):
Thanks a lot for addressing the comments @MarkRich !Looks good, will give it try now and then merge soon :)

patil-suraj commented on examples/community/README.md:
```diff
@@ -371,3 +371,55 @@ for i in range(4):
+device = th.device('cpu' if not has_cuda else 'cuda')
+
+pipe = DiffusionPipeline.from_pretrained(
+    "CompVis/stable-diffusion-v1-4",
```
with safety checker it's returning black images, let's pass None here so the example will work.

```suggestion
    "CompVis/stable-diffusion-v1-4",
    safety_checker=None,
```

patil-suraj reviewed (commented):
The example works great, I just two more comments, then it should be good to merge :)

patrickvonplaten commented on examples/community/README.md:
```diff
@@ -373,6 +372,55 @@ for i in range(4):
+    "CompVis/stable-diffusion-v1-4",
+    use_auth_token=True,
+    custom_pipeline="imagic_stable_diffusion",
+    scheduler = DDIMScheduler(beta_start=0.00085, beta_end=0.012, beta_schedule="scaled_linear", clip_sample=False, set_alpha_to_one=False)
```
```suggestion
    scheduler = DDIMScheduler.from_pretrained("CompVis/stable-diffusion-v1-4", subfolder="scheduler")
```

MarkRich replied:
Hmm `DDIMScheduler.from_pretrained` yields an error (that from_pretrained does not exist) and these parameters specific parameters are actually necessary for the algorithm to produce the expected results.

patrickvonplaten reviewed (approved):
Looks good! Thanks a lot :-) 
@patil-suraj feel free to merge whenever

MarkRich commented:
Updated with comments from the review; should be good to go!

patil-suraj commented:
Thanks a lot @MarkRich ! The tests failures are unrelated, merging!

zhongyi-zhou commented:
@MarkRich Thanks for the amazing code! 
I have one question on this line:
https://github.com/huggingface/diffusers/blob/f9cfb5ab8a632e8bcb96e0a7f0a49a11744f4ba5/examples/community/imagic_stable_diffusion.py#L312

Why it is ```text_embeddings``` instead of ```text_embeddings_orig```?

According to the Imagic paper, in the "model fine-tuning" stage, it says 
- "We fine-tune them with the same reconstruction loss, but conditioned on $e_{tgt}$, as $e_{opt}$ is optimized for the base model only" (Page 5; the second column)
![P 3M N LB(2R}$MPUMVKC9](https://user-images.githubusercontent.com/33302970/224934219-2d00fb90-2338-43b6-9b4d-feaca9c4828f.png)

From my understanding $e_{tgt}$ is ```text_embeddings_orig``` and $e_{opt}$ is ```text_embeddings``` in this code.
In practice, these two values should be very similar, and probably that's the reason why the code still work amazingly well.

Please correct me if I am wrong. Thanks!

shiranzada commented:
Thanks @zhongyi-zhou, we condition on e_tgt during finetuning **only for the super resolution models of Imagen**. This part is not relevant for Stable Diffusion.

For the base model (Imagen-64 or LDM) we condition **on e_opt** during finetuning rather than on e_tgt
